File: scripts/import.mjs
#!/usr/bin/env node
/**
 * Parse comic_collection_tracker.xlsx + Ebay_Purchase_History_*.xlsx
 * and generate src/data/collection.json.
 *
 * Usage: node scripts/import.mjs
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync, readdirSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import XLSX from "xlsx";

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_DIR = join(__dirname, "..", "data");
const OUTPUT = join(__dirname, "..", "src", "data", "collection.json");

function slugify(name) {
  return name.toLowerCase().trim().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "");
}

function readSheet(filePath, sheetName) {
  const wb = XLSX.readFile(filePath);
  const ws = wb.Sheets[sheetName || wb.SheetNames[0]];
  return XLSX.utils.sheet_to_json(ws, { defval: null });
}

function main() {
  const trackerPath = join(DATA_DIR, "comic_collection_tracker.xlsx");

  // --- All Purchases ---
  const purchases = readSheet(trackerPath, "All Purchases");
  const coverA = {};   // series -> Set<vol>
  const allVols = {};   // series -> Set<vol>

  for (const row of purchases) {
    const series = row["Series"];
    const vol = row["Volume #"];
    const cover = row["Cover"] || "";
    if (!series || vol == null) continue;
    const v = Number(vol);
    if (!allVols[series]) allVols[series] = new Set();
    allVols[series].add(v);
    if (cover.toUpperCase().includes("CVR A")) {
      if (!coverA[series]) coverA[series] = new Set();
      coverA[series].add(v);
    }
  }

  // --- Collection Summary (for max volume) ---
  const summary = readSheet(trackerPath, "Collection Summary");
  const summaryMax = {};
  for (const row of summary) {
    const name = row["Series"];
    const volsStr = String(row["Volumes Owned"] || "");
    const nums = [...volsStr.matchAll(/\d+/g)].map((m) => Number(m[0]));
    if (name && nums.length) summaryMax[name] = Math.max(...nums);
  }

  // --- item_name -> { series, vol, cover } mapping ---
  const itemInfo = {};
  for (const row of purchases) {
    const name = (row["Item Name"] || "").trim();
    if (row["Series"] && row["Volume #"] != null && name) {
      itemInfo[name] = {
        series: row["Series"],
        vol: Number(row["Volume #"]),
        cover: row["Cover"] || "",
      };
    }
  }

  // --- Extract images from eBay files ---
  const seriesImages = {};
  // issueCovers: series -> vol -> [ { cover, imageUrl } ]
  const issueCovers = {};

  const ebayFiles = readdirSync(DATA_DIR).filter(
    (f) => f.startsWith("Ebay_Purchase_History_") && f.endsWith(".xlsx")
  );

  // Track seen cover+vol combos to avoid duplicates
  const seenCovers = new Set();

  for (const file of ebayFiles) {
    const rows = readSheet(join(DATA_DIR, file));
    for (const row of rows) {
      const itemName = (row["ItemName"] || "").trim();
      const imgUrl = row["Image URL"];
      if (!imgUrl || !itemName) continue;
      const info = itemInfo[itemName];
      if (!info) continue;
      const isCoverA = info.cover.toUpperCase().includes("CVR A");

      if (!seriesImages[info.series] || isCoverA) {
        seriesImages[info.series] = imgUrl;
      }

      // Collect all covers per issue
      const key = `${info.series}|${info.vol}|${info.cover}`;
      if (!seenCovers.has(key)) {
        seenCovers.add(key);
        if (!issueCovers[info.series]) issueCovers[info.series] = {};
        if (!issueCovers[info.series][info.vol]) issueCovers[info.series][info.vol] = [];
        issueCovers[info.series][info.vol].push({
          cover: info.cover,
          imageUrl: imgUrl,
        });
      }
    }
  }

  // Sort covers per issue: Cover A first, then alphabetically
  for (const series of Object.values(issueCovers)) {
    for (const vol of Object.keys(series)) {
      series[vol].sort((a, b) => {
        const aIsA = a.cover.toUpperCase().includes("CVR A") ? 0 : 1;
        const bIsA = b.cover.toUpperCase().includes("CVR A") ? 0 : 1;
        if (aIsA !== bIsA) return aIsA - bIsA;
        return a.cover.localeCompare(b.cover);
      });
    }
  }

  // --- Build output ---
  const seriesList = Object.keys(allVols)
    .sort()
    .map((name) => {
      const vols = [...allVols[name]].sort((a, b) => a - b);
      const maxVol = summaryMax[name] || Math.max(...vols);
      const ownedA = coverA[name] ? [...coverA[name]].sort((a, b) => a - b) : [];
      const ownedOther = [...allVols[name]]
        .filter((v) => !coverA[name]?.has(v))
        .sort((a, b) => a - b);

      const covers = issueCovers[name] || {};
      const issueCoversOut = {};
      for (const [v, arr] of Object.entries(covers)) {
        issueCoversOut[String(v)] = arr;
      }

      return {
        id: slugify(name),
        name,
        publisher: "DC",
        totalIssues: maxVol,
        ownedCoverA: ownedA,
        ownedOther: ownedOther,
        imageUrl: seriesImages[name] || "",
        issueCovers: issueCoversOut,
      };
    });

  mkdirSync(dirname(OUTPUT), { recursive: true });
  writeFileSync(OUTPUT, JSON.stringify({ series: seriesList }, null, 2));
  console.log(`Wrote ${seriesList.length} series to ${OUTPUT}`);
}

main();

File: scripts/import.py
#!/usr/bin/env python3
"""Parse comic_collection_tracker.xlsx and generate src/data/collection.json."""

import json
import os
import re
from collections import defaultdict

import openpyxl

XLSX_PATH = os.path.expanduser("~/Downloads/comic_collection_tracker.xlsx")
EBAY_FILES = [
    os.path.expanduser(f"~/Downloads/Ebay_Purchase_History_{y}.xlsx")
    for y in [2024, 2025, 2026]
]
OUTPUT_PATH = os.path.join(os.path.dirname(__file__), "..", "src", "data", "collection.json")


def slugify(name: str) -> str:
    s = name.lower().strip()
    s = re.sub(r"[^a-z0-9]+", "-", s)
    return s.strip("-")


def main():
    wb = openpyxl.load_workbook(XLSX_PATH, read_only=True)

    # --- Parse "All Purchases" sheet ---
    ws = wb["All Purchases"]
    rows = list(ws.iter_rows(values_only=True))
    header = rows[0]
    data = rows[1:]

    cover_a: dict[str, set[int]] = defaultdict(set)
    all_volumes: dict[str, set[int]] = defaultdict(set)

    for row in data:
        series = row[1]
        vol = row[2]
        cover = row[3]
        if series is None or vol is None:
            continue
        vol = int(vol)
        all_volumes[series].add(vol)
        if cover and "CVR A" in str(cover).upper():
            cover_a[series].add(vol)

    # --- Parse "Collection Summary" for total issues range ---
    ws2 = wb["Collection Summary"]
    summary_rows = list(ws2.iter_rows(values_only=True))
    summary_header = summary_rows[0]
    summary_data = summary_rows[1:]

    # Build max volume per series from summary "Volumes Owned" column
    summary_max: dict[str, int] = {}
    for row in summary_data:
        series_name = row[0]
        if series_name is None:
            continue
        # Volumes Owned is like "1-17" or "1, 15" or "1"
        vols_str = str(row[1])
        nums = [int(x) for x in re.findall(r"\d+", vols_str)]
        if nums:
            summary_max[series_name] = max(nums)

    wb.close()

    # --- Build item_name -> (series, vol, cover) mapping for image lookup ---
    item_to_info: dict[str, tuple[str, int, str]] = {}
    for row in data:
        if row[1] and row[2] and row[4]:
            item_to_info[str(row[4]).strip()] = (row[1], int(row[2]), row[3] or "")

    # --- Extract image URLs from eBay files ---
    # Per-series: one image (prefer Cover A)
    # Per-issue: series -> vol -> {cover, url} (prefer Cover A)
    series_images: dict[str, str] = {}
    issue_images: dict[str, dict[int, str]] = defaultdict(dict)  # series -> vol -> best url

    for ebay_file in EBAY_FILES:
        if not os.path.exists(ebay_file):
            continue
        ewb = openpyxl.load_workbook(ebay_file, read_only=True)
        ews = ewb[ewb.sheetnames[0]]
        for row in list(ews.iter_rows(values_only=True))[1:]:
            item_name = (row[4] or "").strip()
            img_url = row[11]
            if not img_url or not item_name:
                continue
            info = item_to_info.get(item_name)
            if not info:
                continue
            series_name, vol, cover = info
            is_cover_a = "CVR A" in cover.upper()

            # Series-level image
            if series_name not in series_images or is_cover_a:
                series_images[series_name] = img_url

            # Issue-level image (prefer Cover A)
            if vol not in issue_images[series_name] or is_cover_a:
                issue_images[series_name][vol] = img_url
        ewb.close()

    # --- Build output ---
    series_list = []
    for name in sorted(all_volumes.keys()):
        vols = sorted(all_volumes[name])
        max_vol = summary_max.get(name, max(vols))
        owned_a = sorted(cover_a.get(name, set()))
        owned_other = sorted(all_volumes[name] - cover_a.get(name, set()))

        series_list.append({
            "id": slugify(name),
            "name": name,
            "publisher": "DC",
            "totalIssues": max_vol,
            "ownedCoverA": owned_a,
            "ownedOther": owned_other,
            "imageUrl": series_images.get(name, ""),
            "issueImages": {str(v): url for v, url in sorted(issue_images.get(name, {}).items())},
        })

    os.makedirs(os.path.dirname(OUTPUT_PATH), exist_ok=True)
    with open(OUTPUT_PATH, "w") as f:
        json.dump({"series": series_list}, f, indent=2)

    print(f"Wrote {len(series_list)} series to {OUTPUT_PATH}")


if __name__ == "__main__":
    main()
